### PR TITLE
Spark: IN clause on system function is not pushed down

### PR DESCRIPTION
**Pull request by tmnd1991**
PR to verify bug reported in issue #9191.

With some guidance I'm open to work on the fix too.

### CODE REVIEW

**Comment by tmnd1991** [comment]
Can I have a feedback on this one @ConeyLiu, too? Thanks a lot

File: spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceStaticInvoke.scala [hunk 1]
```diff
@@ -40,14 +37,20 @@ import org.apache.spark.sql.types.StructType
...
       case filter @ Filter(condition, _) =>
-        val newCondition = condition.transformWithPruning(_.containsPattern(BINARY_COMPARISON)) {
+        val newCondition = condition.transformWithPruning(_.containsAnyPattern(BINARY_COMPARISON, IN, INSET)) {
           case c @ BinaryComparison(left: StaticInvoke, right) if canReplace(left) && right.foldable =>
             c.withNewChildren(Seq(replaceStaticInvoke(left), right))
 
           case c @ BinaryComparison(left, right: StaticInvoke) if canReplace(right) && left.foldable =>
             c.withNewChildren(Seq(left, replaceStaticInvoke(right)))
+
+          case in @ In(s: StaticInvoke, _) =>
```
Comment (ConeyLiu) [inline spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceStaticInvoke.scala:49]: why not check if `s` can be replaced or not?

**Comment by ConeyLiu** [comment]
@tmnd1991 thanks for your contribution. I think this is a valid fix.

File: spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceStaticInvoke.scala (hunk 1 above)
Comment (tmnd1991) [reply spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceStaticInvoke.scala]: my bad, I fixed it!

**Comment by tmnd1991** [comment]
cc @nastra @dramaticlly @advancedxy for review
thanks

File: spark/v3.4/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestSystemFunctionPushDownDQL.java [hunk 2]
```diff
@@ -224,6 +226,35 @@ private void testBucketLongFunction(boolean partitioned) {
     Assertions.assertThat(actual.size()).isEqualTo(5);
   }
 
+  @Test
+  public void testBucketLongFunctionInClauseOnUnpartitionedTable() {
```
[the remark this answers is not in the document]
Comment (tmnd1991) [reply spark/v3.4/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestSystemFunctionPushDownDQL.java:230]: yes, that was the idea, let's focus on 3.5 and when we reach consensus on that I back port it to 3.4

[the remark this answers is not in the document]
Comment (tmnd1991) [reply spark/v3.4/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestSystemFunctionPushDownDQL.java]: yup

File: spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceStaticInvoke.scala [hunk 3]
```diff
@@ -40,14 +37,20 @@ import org.apache.spark.sql.types.StructType
...
       case filter @ Filter(condition, _) =>
-        val newCondition = condition.transformWithPruning(_.containsPattern(BINARY_COMPARISON)) {
+        val newCondition = condition.transformWithPruning(_.containsAnyPattern(BINARY_COMPARISON, IN, INSET)) {
           case c @ BinaryComparison(left: StaticInvoke, right) if canReplace(left) && right.foldable =>
             c.withNewChildren(Seq(replaceStaticInvoke(left), right))
 
           case c @ BinaryComparison(left, right: StaticInvoke) if canReplace(right) && left.foldable =>
             c.withNewChildren(Seq(left, replaceStaticInvoke(right)))
+
+          case in @ In(value: StaticInvoke, _) if canReplace(value) =>
```
[the remark this answers is not in the document]
Comment (tmnd1991) [reply spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceStaticInvoke.scala]: done as suggested

File: spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceStaticInvoke.scala [hunk 4]
```diff
@@ -40,14 +37,23 @@ import org.apache.spark.sql.types.StructType
...
-        val newCondition = condition.transformWithPruning(_.containsPattern(BINARY_COMPARISON)) {
+        val newCondition = condition.transformWithPruning(_.containsAnyPattern(BINARY_COMPARISON, IN, INSET)) {
           case c @ BinaryComparison(left: StaticInvoke, right) if canReplace(left) && right.foldable =>
             c.withNewChildren(Seq(replaceStaticInvoke(left), right))
 
           case c @ BinaryComparison(left, right: StaticInvoke) if canReplace(right) && left.foldable =>
             c.withNewChildren(Seq(left, replaceStaticInvoke(right)))
+
+          case in @ In(systemFunction: StaticInvoke, values)
+            if canReplace(systemFunction) && values.forall(_.foldable) =>
```
Comment (wypoon) [inline spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceStaticInvoke.scala:50]: Nit: Indent the `if` one more level for readability of this block?

File: spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceStaticInvoke.scala [hunk 5]
```diff
@@ -40,14 +37,23 @@ import org.apache.spark.sql.types.StructType
...
+            in.copy(value = replaceStaticInvoke(systemFunction))
+
+          case in@InSet(systemFunction: StaticInvoke, _) if canReplace(systemFunction) =>
```
Comment (wypoon) [inline spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceStaticInvoke.scala:53]: Nit: `in @ InSet` to be consistent with the style.

File: spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestSystemFunctionPushDownDQL.java [hunk 6]
```diff
@@ -213,17 +198,77 @@ private void testBucketLongFunction(boolean partitioned) {
...
+  }
+
+  private void testBucketLongFunctionIsNotReplacedWhenArgumentsAreNotLiterals() {
+    List<Integer> range = IntStream.range(0, 3).boxed().collect(Collectors.toList());
+    String rangeAsSql =
+        range.stream().map(x -> Integer.toString(x)).collect(Collectors.joining(", "));
+    String query =
+        String.format(
+            "SELECT * FROM %s WHERE system.bucket(5, id) IN (system.bucket(5, id), 1) ORDER BY id",
+            tableName, rangeAsSql);
```
Comment (wypoon) [inline spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestSystemFunctionPushDownDQL.java:259]: `range` and `rangeAsSql` are not needed, since the `IN` clause here doesn't use `rangeAsSql`.

**Review by wypoon — COMMENTED** [review]
The fix seems to make sense.
Just one comment about a test and some nits (also, please note that the Iceberg project doesn't use wildcard imports, so you should put back all the explicit imports plus additional ones needed).

**Comment by tmnd1991** [comment]
Thanks @wypoon for the review, I addressed your concerns

File: spark/v3.4/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceStaticInvoke.scala [hunk 7]
```diff
@@ -56,7 +69,6 @@ object ReplaceStaticInvoke extends Rule[LogicalPlan] {
           filter.copy(condition = newCondition)
         }
   }
-
```
Comment (wypoon) [inline spark/v3.4/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceStaticInvoke.scala:59]: Unnecessary (and undesirable) whitespace change.

File: spark/v3.4/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestSystemFunctionPushDownDQL.java [hunk 8]
```diff
@@ -213,17 +216,75 @@ private void testBucketLongFunction(boolean partitioned) {
...
+  }
+
+  @Test
+  public void testBucketLongFunctionIsNotReplacedWhenArgumentsAreNotLiteralsOnUnpartitionedTable() {
+    createUnpartitionedTable(spark, tableName);
+    testBucketLongFunctionIsNotReplacedWhenArgumentsAreNotLiterals();
+  }
+
+  private void testBucketLongFunctionIsNotReplacedWhenArgumentsAreNotLiterals() {
+    List<Integer> range = IntStream.range(0, 3).boxed().collect(Collectors.toList());
```
Comment (wypoon) [inline spark/v3.4/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestSystemFunctionPushDownDQL.java:271]: `range` is not used. (I pointed this out before.)

File: spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceStaticInvoke.scala [hunk 9]
```diff
@@ -56,7 +69,6 @@ object ReplaceStaticInvoke extends Rule[LogicalPlan] {
           filter.copy(condition = newCondition)
         }
   }
-
```
Comment (wypoon) [inline spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceStaticInvoke.scala:59]: Unnecessary whitespace change.

File: spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestSystemFunctionPushDownDQL.java [hunk 10]
```diff
@@ -213,17 +216,75 @@ private void testBucketLongFunction(boolean partitioned) {
...
+  }
+
+  @Test
+  public void testBucketLongFunctionIsNotReplacedWhenArgumentsAreNotLiteralsOnUnpartitionedTable() {
+    createUnpartitionedTable(spark, tableName);
+    testBucketLongFunctionIsNotReplacedWhenArgumentsAreNotLiterals();
+  }
+
+  private void testBucketLongFunctionIsNotReplacedWhenArgumentsAreNotLiterals() {
+    List<Integer> range = IntStream.range(0, 3).boxed().collect(Collectors.toList());
```
Comment (wypoon) [inline spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestSystemFunctionPushDownDQL.java:271]: Not used.

**Review by wypoon — COMMENTED** [review]
LGTM otherwise.

**Comment by wypoon** [comment]
@rdblue @szehon-ho @RussellSpitzer can you please review this? Also, can you please enable the CI for this PR?

**Comment by tmnd1991** [comment]
any chance to have this reviewed?

**Comment by nastra** [comment]
@aokolnychyi could you take a look at this please?

**Comment by nastra** [comment]
just FYI, looks like there's a related PR that @aokolnychyi opened for this: https://github.com/apache/iceberg/pull/9873

**Comment by tmnd1991** [comment]
> just FYI, looks like there's a related PR that @aokolnychyi opened for this: #9873

thanks, watching that 🙏

**Comment by tmnd1991** [comment]
@aokolnychyi I see you fixed part of this in https://github.com/apache/iceberg/pull/9873 but Spark 3.4 looks stil bugged on main. Do you have any time to give me a feedback on this?
